nss_simpleauthd/src/lib.rs

```diff
@@ -1,48 +1,47 @@
-use authd::types::ToNSS;
-use futures::executor::block_on;
-use libc::c_int;
-use libnss::group::{CGroup, Group, GroupHooks};
-use libnss::interop::{Iterator, Response};
-use libnss::passwd::{CPasswd, Passwd};
-use std::ffi::CStr;
-use std::str;
-use std::sync::{Arc, Mutex, MutexGuard};
+mod config;
+mod ffi;
+mod group;
+mod passwd;
+mod shadow;
+
+use std::sync::{Arc, Mutex};
 use std::time::{Duration, Instant};
-use tarpc::context;
 use tokio::runtime::{self, Runtime};
 use tokio::time::sleep_until;
+use trust_dns_resolver::TokioAsyncResolver;
+
+use crate::config::CFG;
+
+lazy_static::lazy_static! {
+    static ref RPC: Mutex<ClientAccessControl> = Mutex::new(ClientAccessControl::default());
+    static ref RT: Runtime = runtime::Builder::new_multi_thread().worker_threads(2).enable_io().enable_time().build().expect("could not initialize tokio runtime");
+}
+
+struct AuthdPasswd;
+struct AuthdShadow;
+struct AuthdGroup;
 
 #[derive(Default)]
 struct ClientAccessControl {
     client: Arc<Mutex<Option<authd::rpc::AuthdClient>>>,
     latest_ts: Arc<Mutex<Option<Instant>>>,
 }
 
-#[derive(serde::Deserialize)]
-struct NssConfig {
-    host: authd::SocketName,
-    cert: String,
-}
-
 impl ClientAccessControl {
     fn with_client<O>(&mut self, f: impl FnOnce(&mut authd::rpc::AuthdClient) -> O) -> O {
-        use trust_dns_resolver::TokioAsyncResolver;
         let _guard = RT.enter();
         let mut lts = self.latest_ts.lock().unwrap();
         *lts = Some(std::time::Instant::now() + Duration::from_secs(30));
-        let cl = self.client.clone();
+
+        let client = self.client.clone();
         let latest_ts = self.latest_ts.clone();
         tokio::spawn(async move {
             loop {
                 let dur = latest_ts.lock().unwrap().unwrap_or(Instant::now()).into();
                 sleep_until(dur).await;
                 // make sure it wasn't moved forward while we were sleeping
                 if latest_ts.lock().unwrap().unwrap_or(Instant::now()) < Instant::now() {
-                    *cl.lock().unwrap() = None;
-                    #[cfg(debug_assertions)]
-                    eprintln!(
-                        "nss_simpleauthd: ClientAccessControl: client timed out, closing connection."
```

Comment:
what's wrong with this?

Reply:
Good catch, I was getting ahead of myself. I'm pretty sure #2 (NSS modules shouldn't crash) but I was thinking that NSS modules shouldn't talk. I'll bring that back.